feat: add write blog script

BOLT04 opened the pull request:
**Description**

- Little demo 🙂 : https://recordit.co/S0oaxPbmP1
- I removed the authors and layout prompts of the initial script, but we could add them too, just let me know 👍 

**Related issue(s)**
See also #90

BOLT04 committed:
docs: add section to write a new blog post

#90

BOLT04 committed:
chore: add write blog script and dependencies

BOLT04 committed:
feat: add compose.js script

#90

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

BOLT04 committed:
docs: add example embeds and more info

boyney123 commented:
Very cool @BOLT04 👏 like it!

boyney123 commented on scripts/compose.js:
```diff
@@ -0,0 +1,147 @@
+      name: 'extension',
+      message: 'Choose post extension:',
+      type: 'list',
+      choices: ['md'],
```
`Choose post extension`, is there other extensions they can use? If it's just one (which is `md`) shall we remove this question and assume it's always `md`? Unless I'm missing something?

BOLT04 committed:
fix: only allow md extension on compose.js

BOLT04 committed:
fix: back ticks and add more examples

quetzalliwrites commented:
THIS IS SOOOO COOL!!! 😎

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

BOLT04 committed:
chore: update package-lock.json

BOLT04 commented:
Hi @alequetzalli and @derberg, I noticed some conflicts so I updated the branch from master, but the checks still failed because it throws this error `ERR_REQUIRE_ESM`.
After some digging it seems the PR #553 introduced some breaking changes since the package `node-fetch` is ESM only and we were using `const fetch = require('node-fetch')`.

Should we downgrade to v2, make a workaround in this PR or simply use `import` statements on our scripts and fix the errors in this PR?

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

BOLT04 commented:
> Hi @alequetzalli and @derberg, I noticed some conflicts so I updated the branch from master, but the checks still failed because it throws this error `ERR_REQUIRE_ESM`. After some digging it seems the PR #553 introduced some breaking changes since the package `node-fetch` is ESM only and we were using `const fetch = require('node-fetch')`.
> 
> Should we downgrade to v2, make a workaround in this PR or simply use `import` statements on our scripts and fix the errors in this PR?

fixed in #565 🙂

derberg reviewed (commented):
@BOLT04 I cloned it locally, ran `npm install` and this is what I get when running the script:

```
node ./scripts/compose.js
internal/modules/cjs/loader.js:968
  throw err;
  ^

Error: Cannot find module 'figures'
Require stack:
- /Users/wookiee/sources/website-asyncapi/node_modules/inquirer/lib/objects/separator.js
- /Users/wookiee/sources/website-asyncapi/node_modules/inquirer/lib/inquirer.js
- /Users/wookiee/sources/website-asyncapi/scripts/compose.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:965:15)
    at Function.Module._load (internal/modules/cjs/loader.js:841:27)
    at Module.require (internal/modules/cjs/loader.js:1025:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (/Users/wookiee/sources/website-asyncapi/node_modules/inquirer/lib/objects/separator.js:3:17)
    at Module._compile (internal/modules/cjs/loader.js:1137:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
    at Module.load (internal/modules/cjs/loader.js:985:32)
    at Function.Module._load (internal/modules/cjs/loader.js:878:14)
    at Module.require (internal/modules/cjs/loader.js:1025:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/wookiee/sources/website-asyncapi/node_modules/inquirer/lib/objects/separator.js',
    '/Users/wookiee/sources/website-asyncapi/node_modules/inquirer/lib/inquirer.js',
    '/Users/wookiee/sources/website-asyncapi/scripts/compose.js'
  ]
}
```

I tried node 16 and 12, and get the same. Fun fact, when I take this code outside website, in a standalone project, it works well 🤔 

I checked and `figures` is really not getting installed. I tried to remove all `node-modules` and then install all from scratch, it did not help.

When I add it manually I get
```
$ node ./scripts/compose.js
node:internal/modules/cjs/loader:1112
      throw new ERR_REQUIRE_ESM(filename, parentPath, packageJsonPath);
      ^

Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /Users/wookiee/sources/website-asyncapi/node_modules/figures/index.js
require() of ES modules is not supported.
require() of /Users/wookiee/sources/website-asyncapi/node_modules/figures/index.js from /Users/wookiee/sources/website-asyncapi/node_modules/inquirer/lib/objects/separator.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /Users/wookiee/sources/website-asyncapi/node_modules/figures/package.json.

    at new NodeError (node:internal/errors:363:5)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1112:13)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:816:12)
    at Module.require (node:internal/modules/cjs/loader:999:19)
    at require (node:internal/modules/cjs/helpers:93:18)
    at Object.<anonymous> (/Users/wookiee/sources/website-asyncapi/node_modules/inquirer/lib/objects/separator.js:3:17)
    at Module._compile (node:internal/modules/cjs/loader:1095:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1124:10)
    at Module.load (node:internal/modules/cjs/loader:975:32) {
  code: 'ERR_REQUIRE_ESM'
}
```

tbh I have no idea what is going on here 🤔

Commit:
Merge branch 'master' into feat/90-add-write-blog-script

quetzalliwrites commented:
> 
> fixed in #565 🙂

Were you able to fully figure it out ok, @BOLT04? 🙂 If not, I still think asking for help from folks like @derberg, @magicmatatjahu or other `/website` repo owners might be best suited to help out 💚

BOLT04 commented:
> > fixed in #565 🙂
> 
> Were you able to fully figure it out ok, @BOLT04? 🙂 If not, I still think asking for help from folks like @derberg, @magicmatatjahu or other `/website` repo owners might be best suited to help out 💚

yes it's solved now, thanks @alequetzalli 🙂.

 About the error: `Error: Cannot find module 'figures'` @derberg, maybe we could just explicitly add `figures` as a dependency on this project. That should cover this edge case where the module doesn't get installed sometimes, wdyt?

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

derberg commented:
@BOLT04 yeah, I did try to add it manually and got the next error that is in my previous comment, that is the problem 😞 
I'll try to investigate as this is really puzzling me

derberg commented:
@BOLT04 ok, know how to solve it, got hint from https://stackoverflow.com/questions/65974337/import-es-module-in-next-js-err-require-esm although I don't know why it matter as you have here a script that is triggered independently from next.js

Anyway, I updated `next` in package.json to latest `12.0.9` and it works! also tried to build and run dev locally and nothing was failing. So yeah just update `next` and push to PR and in the meantime I will review functionality

derberg commented on scripts/compose.js:
```diff
@@ -0,0 +1,150 @@
+  type: ${answers.type}
+  tags: [${answers.tags ? tags : ''}]
+  cover: /img/posts/may-2021-at-asyncapi/cover.webp
+  draft: ${answers.draft === 'yes' ? true : false}
```
you can remove it, we do not support it here, like nothing will happen if you have it in front matter or not

BOLT04 replied:
what about the `type`, should I remove it too?

derberg replied:
no no, type is used, is important. It is consumed in UI as one of filters, and also shows as a kind of article type name

![Screenshot 2022-01-27 at 21 11 01](https://user-images.githubusercontent.com/6995927/151436099-13c33636-4750-43b5-a1ce-bffcd42c3588.png)

BOLT04 replied:
so the one I should remove is just `draft`?

BOLT04 replied:
@derberg I removed the `draft` option 👍

derberg commented on scripts/compose.js:
```diff
@@ -0,0 +1,150 @@
+
+  let frontMatter = dedent`---
+  title: ${answers.title ? answers.title : 'Untitled'}
+  date: '${date}'
```
it should be in `2020-12-23T06:00:00+01:00` format and is `2022-01-27`

derberg commented on scripts/compose.js:
```diff
@@ -0,0 +1,150 @@
+  Add a cover image, you can change it on the metadata field above called \`cover\`. In case you need some inspiration we recommend https://unsplash.com/.
+  All images should be stored in the \`public/img/posts/\` folder, below is an example of using an image in your post:
+
+  ![Figure 3: APIs give developer teams more independence](/img/posts/openapi-vs-asyncapi-burning-questions/asyncapi-openapi-post_pic-03.webp)
```
I would recommend using dedicated component:
```
<Figure
  src="/img/posts/2020-summary/linkedin-folowers.webp"
  caption="Figure 3: Followers growth in 2020"
/>
```

as it also enforces you think about accessibility because caption will also be used as alt text. You can also mention it

derberg reviewed (changes requested):
just left 3 comments
man, I'm so happy you're adding this stuff!

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

BOLT04 committed:
chore(deps): update next to latest

BOLT04 committed:
fix: date and Figure component

BOLT04 committed:
chore(deps): update inquirer to latest

BOLT04 commented:
@derberg can we upgrade to `next` `12.0.9`? I think the Netlify checks are failing because of it, or not?

derberg commented:
@BOLT04 are they? they should not. I tested locally and all was fine. What was failing recently was uncontrolled upgrade to latest `node-fetch`

BOLT04 commented:
> @BOLT04 are they? they should not. I tested locally and all was fine. What was failing recently was uncontrolled upgrade to latest `node-fetch`

@derberg could it be the node version used in netlify? I saw that Next 12 introduced a breaking change where the minimum node version was increased: "The minimum Node.js version has been bumped from 12.0.0 to 12.22.0, which is the first version of Node.js with native ES modules support."

Do you suggest trying something else? I may test a commit rolling back the `next` update.

BOLT04 committed:
chore(deps): downgrade next to previous version

BOLT04 commented:
@derberg I downgraded to the previous version of next and the script still works. Let me know if it works for you too 🙂

derberg commented:
@BOLT04 what do you mean you downgraded next? I see no changes in `package.json`, and anyway, solution is to upgrade to the latest, not to downgrade to the previous 😄 we should go up with version as by downgrading, especially in case of a core like nextjs, we only mitigate the issue by moving it to the future

BOLT04 commented:
> @BOLT04 what do you mean you downgraded next? I see no changes in `package.json`, and anyway, solution is to upgrade to the latest, not to downgrade to the previous 😄 we should go up with version as by downgrading, especially in case of a core like nextjs, we only mitigate the issue by moving it to the future

@derberg hahaha sorry for the confusion 😅. What I mean is I upgraded `next` to the version you specified in [this comment](https://github.com/asyncapi/website/pull/547#issuecomment-1023276575), this was done in [this commit](https://github.com/asyncapi/website/pull/547/commits/96aff5e0e884d616cf5209ba0c458c0c5efd3742). Then the build broke, so I reverted those changes in [this commit](https://github.com/asyncapi/website/pull/547/commits/7c95d51043a5300ff7607d856a2f3e35bb9c892c).
TL;DR: there are no changes to `next` in this PR 😃

If we do need to upgrade to `next` version `12.0.9`, could you help me fix the Netlify checks that fail when this upgrade is done? For example this build failed previously: https://github.com/asyncapi/website/runs/4971727317

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

BOLT04 committed:
refactor: remove draft input

BOLT04 commented:
@derberg you can review again when you're available 👍

derberg reviewed (dismissed):
Man! forget it, I don't know why but now all works without issues!

Awesome stuff!!!

@alequetzalli do you want to check this small addition to readme + the instruction that @BOLT04 suggested in generated blog post?

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

Commit:
Merge branch 'master' into feat/90-add-write-blog-script

quetzalliwrites commented on README.md:
```diff
@@ -45,6 +45,15 @@ npm run dev
 
+#### Compose a new blog post
+
+To bootstrap a new post use this command:
```
```suggestion
To bootstrap a new post, run this command:
```

quetzalliwrites reviewed (commented):
Just one tiny review suggestion, @BOLT04 😄👍🏽

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

BOLT04 committed:
docs: add suggestion

Co-authored-by: Alejandra Quetzalli  <alejandra.quetzalli@postman.com>

BOLT04 commented:
@derberg and @alequetzalli you can review again, thanks for the suggestions 👍

quetzalliwrites reviewed (approved):
🚢

magicmatatjahu committed:
Merge branch 'master' into feat/90-add-write-blog-script

magicmatatjahu commented:
@BOLT04 Łukasz is on 2 weeks vacation, so I'm gonna handle that. Am I to understand that we can merge this? :)

BOLT04 commented:
Hi @magicmatatjahu, yes you can merge it 🙂. If you can also review this one: https://github.com/asyncapi/.github/pull/118, these 2 PRs go hand in hand 👍

magicmatatjahu committed:
Merge branch 'master' into feat/90-add-write-blog-script

magicmatatjahu commented:
@BOLT04 Ok, I wrote one simple suggestion in the https://github.com/asyncapi/.github/pull/118 but I'm not the maintainer of mentioned repo so we need to wait for one of codeowners and then I will merge that PR :)

magicmatatjahu committed:
Merge branch 'master' into feat/90-add-write-blog-script

BOLT04 committed:
Merge branch 'master' into feat/90-add-write-blog-script

derberg committed:
Merge branch 'master' into feat/90-add-write-blog-script

derberg commented:
/rtm

derberg commented:
@all-contributors please add BOLT04 for code,docs

derberg commented:
@all-contributors please add BOLT04 for code, docs

derberg commented:
@all-contributors please add BOLT04 for code, docs

derberg commented:
finally!
